Add Callbacks Section to Logging & Monitoring

john-jac opened the pull request:
Airflow users often ask for an easy way to monitor the success or failure of tasks and/or DAGs.  The use of callbacks, such as on_success_callback, helps users monitor DAG runs in an easily configurable way.

This document provides an overview of available callbacks, linked to the relevant section in the Tasks document.  It also provides a simple example of calling a function on any task failure in the DAG, as well as a separate function that is called upon the success of the last task in the DAG.

john-jac committed:
Add Callbacks Section to Logging & Monitoring

Airflow users often ask for an easy way to monitor the success or failure of tasks and/or DAGs.  The use of callbacks, such as on_success_callback, helps users monitor DAG runs in an easily configurable way.

potiuk commented on docs/apache-airflow/logging-monitoring/callbacks.rst:
```diff
@@ -0,0 +1,75 @@
+
+A valuable component of logging and monitoring is the use of task callbacks to act upon changes in state of a given task, or across all tasks in a given DAG.
+For example, you may wish to alert when certain tasks have failed, or have the last task in your DAG invoke a callback when it succeeds.
+
```
Really Nice! I think we should however add better description of some 'unexpected' callback behaviours. Namely, that callbacks are not executed when the task status is manually set to failed/succeeded from teh UI/CLI. We've had a number of people complaining about it, but this is really an architectural decision (we cannot run callback in those cases because we have no worker which would run those callbacks at the moment we set the status).

 It woudl be great to add this here.

john-jac replied:
Thanks Jarek!  I've added a note indicating that callback functions are not invoked in the case of state change from UI or CLI.

john-jac committed:
Added note regarding limits of callback functions

Added following note: Callback functions are only invoked when the task state changes due to execution by a worker. As such, task changes set by the command line interface (CLI) or user interface (UI) do not execute callback functions.
